Bug 1852553 - Record click events on pages via glean.js

abhi-agg opened the pull request:
Fixes a part of https://bugzilla.mozilla.org/show_bug.cgi?id=1852553

- The click events of the Ping Detail page are missing right now to completely close https://bugzilla.mozilla.org/show_bug.cgi?id=1852553

Additonally, small refactoring: Encapsulated page load event code in telemetry.js

abhi-agg commented:
@rosahbruno Created this PR on top of the [existing one](https://github.com/mozilla/debug-ping-view/pull/85) in case you get a chance to review these changes before we merge [existing one](https://github.com/mozilla/debug-ping-view/pull/85). Once the [existing one](https://github.com/mozilla/debug-ping-view/pull/85) gets merged to `main`, I can create this one against `main` branch too.

rosahbruno reviewed (commented):
I am not sure if it is directly from a change in this PR or not, but when I was starting the project I encountered a bug where the app was in an infinite loop of sending pings which ended up locking the app.

This happened after running the app from a clean install. The app was unresponsive and this is what I saw in the console.

<img width="712" alt="Screenshot 2023-09-18 at 6 12 04 PM" src="https://github.com/mozilla/debug-ping-view/assets/24759139/7bbff352-1474-47ad-8d40-782a57e1156d">

rosahbruno commented on src/components/ActiveClients.js:
```diff
@@ -133,7 +133,7 @@ const ActiveClients = () => {
                 <tr key={key}>
                   <td>
-                    <Link className='text-decoration-none' to={`/pings/${debugId}`}>
+                    <Link className='text-decoration-none' to={`/pings/${debugId}`} onClick={recordClick('Debug ID')}>
```
I think this is causing the issue I mentioned in the previous comment. This function is being immediately invoked on render, which is causing every single row in `ActiveClients` to call `recordClick()` on mount.

This needs to be `onClick={() => recordClick('Debug ID')}`. In the rest of the places that we are using `recordClick` we are doing so inside of a callback function, since this isn't inside of a callback you either need to make this a function that calls a function (not necessary here) or just wrap it in a callback.

abhi-agg replied:
Thanks for catching this bug. Yeah, it was a miss. I fixed it in the next commit.

rosahbruno reviewed (changes requested):
I found what was causing the issue I mentioned above

abhi-agg commented:
@rosahbruno I fixed the issue that you caught.

Additionally, I changed the target branch of this PR to `main` and rebased this PR on `main`. From the perspective of the code to be reviewed, you will not see any change because of this action.

rosahbruno reviewed (changes requested):
There are a few places I think we should add more click events

1. The theme button in the top right - curious to see if this was worth the effort
2. The bug button in the top right - How often do people click this? They might not always file a bug, but they might see a list of bugs and see something they are waiting to have fixed.
3. The help button in the top right - Our help button doesn't offer much assistance, but it might be nice to see how often people feel confused or like they need more info.
4. The sign out button in the top right - Could give us an idea if there are performance issues or login issues that make people think signing in again might fix it.
5. The dynamic `summary` element in the `PingSection` component - We already have click events for these for `Events` and `Metrics` we should have the same here.
6. Breadcrumbs - we don't need to pass any specific key, but just seeing how often those are used could be nice.
7. The `ReadMore` component - this is where users can expand/collapse the ping on the pings screen. It could be useful to see how many people use this as opposed to looking at the ping individually.
8. Inside of `handleLineNumberClick` - gives us an idea of how many people are using the line number specific linking feature

rosahbruno commented on src/components/ReturnToTop.js:
```diff
@@ -23,6 +24,8 @@ const ReturnToTop = () => {
       className='return-to-top btn btn-sm btn-outline-secondary'
       style={{ display: visible ? 'block' : 'none' }}
       onClick={() => {
+        // record this click event
```
nit: probably don't need this comment, but not a deal-breaker

rosahbruno commented on src/components/ShowRawPing/components/ShowRawPing.js:
```diff
@@ -61,7 +61,7 @@ const ShowRawPing = ({ docId }) => {
   /// lifecycle ///
   useEffect(() => {
     // record page load event
```
Same with this, but again not a deal-breaker

rosahbruno reviewed (approved):
Tested and everything looks good. We can add the rest of the click events in a separate PR.

abhi-agg commented:
> There are a few places I think we should add more click events
> 
>     1. The theme button in the top right - curious to see if this was worth the effort
> 
>     2. The bug button in the top right - How often do people click this? They might not always file a bug, but they might see a list of bugs and see something they are waiting to have fixed.
> 
>     3. The help button in the top right - Our help button doesn't offer much assistance, but it might be nice to see how often people feel confused or like they need more info.
> 
>     4. The sign out button in the top right - Could give us an idea if there are performance issues or login issues that make people think signing in again might fix it.
> 
>     5. The dynamic `summary` element in the `PingSection` component - We already have click events for these for `Events` and `Metrics` we should have the same here.
> 
>     6. Breadcrumbs - we don't need to pass any specific key, but just seeing how often those are used could be nice.
> 
>     7. The `ReadMore` component - this is where users can expand/collapse the ping on the pings screen. It could be useful to see how many people use this as opposed to looking at the ping individually.
> 
>     8. Inside of `handleLineNumberClick` - gives us an idea of how many people are using the line number specific linking feature

Hey @rosahbruno Thanks for the feedback. As, unlike page load events, there are quite a number of places that need instrumentation for click events, my plan was to incremently add click events and submit smaller PRs so that:
1. They are easy to review for you compared to 1 big PR
2. I get initial feedback from you sooner by instrumenting only few places

I was thinking of submitting another PR to instrument the rest of the places you mentioned above and not block this one from merging unless you see some issue with the instrumentation code.
